README.md
# Cosinet Package

Cosinet is an R package that aims to bridge the gap between differential
co-expression network analysis and precision medicine by accurately
quantifying the degree of rewiring of co-expression networks at
the level of individual samples. Differential co-expression network analysis
is a valuable approach for understanding the underlying mechanisms of disease
and informing the development of personalized therapies, but current
approaches typically measure an average differential network across multiple
samples, making it difficult to apply in the field of precision medicine.
Cosinet utilizes gene expression data to determine the degree of similarity
between the gene co-expression patterns of a given sample and reference
conditions within the context of a function-specific differential
co-expression network. To obtain differential co-expression network, 
we provide R function for z-score-based differential co-expression analysis. 
Next, we calculate eigenvector centrality to get a ranked gene list, 
which represents the degree of hubness of each gene in the network.
Based on the ranking, one can further perform gene set enrichment analysis
(GSEA) to get enriched pathways, and use the core enrichment genes of a relevant 
pathway to form a function-specific differential co-expression sub-network. 
Finally, Cosinet scores can be calculated based on the specific 
sub-network.


## Installation

To install the cosinet package, you can use the `remotes` package:

``` r
if (!requireNamespace("remotes", quietly=TRUE))
    install.packages("remotes")
remotes::install_github("LanyingWei/cosinet")
```

## Vignette

For more details, please refer to the 
[online vignette](https://htmlpreview.github.io/?https://github.com/LanyingWei/cosinet/blob/main/vignettes/cosinet.html).


NEWS
Changes in version 0.99.0 (2022-11-11)
+ Submitted to Bioconductor


R/data.R
#' Expression data from breast cancer patients
#'
#' This matrix contains a subset of RNA-seq gene expression data from patients
#' with breast cancer. The full data is available in the Gene Expression
#' Omnibus (GEO) repository under accession number GSE96058. The format of the
#' expression value is log2(FPKM + 0.1) and was rounded to one digit to reduce
#' the data size. The corresponding sample information is stored in the
#' data object "samplesBC".
#'
#' @format A matrix with 12,787 rows (genes) and 400 columns (samples).
#' @usage data(exprBC)
#'
"exprBC"

#' Phenotype data from breast cancer patients.
#'
#' This data frame contains a subset of phenotype data from patients with
#' breast cancer. Full data are available in the Gene Expression Omnibus
#' (GEO) repository under accession number GSE96058. The corresponding
#' expression data is stored in the data object "exprBC".
#'
#' @format A data frame containing 11 variables:
#' \describe{
#'  \item{title}{Sample title.}
#'  \item{age_at_diagnosis}{Age at diagnosis.}
#'  \item{chemo_treated}{Whether the patient received adjuvant chemotherapy
#'  (1 for "yes", 0 for "no").}
#'  \item{endocrine_treated}{Whether the patient received adjuvant endocrine
#'  therapy (1 for "yes", 0 for "no").}
#'  \item{er_status}{Estrogen receptor status (1 for "positive", 0 for
#'  "negative").}
#'  \item{her2_status}{Human epidermal growth factor receptor 2 (HER2) status
#'  (1 for "positive", 0 for "negative").}
#'  \item{pgr_status}{Progesterone receptor status (1 for "positive", 0 for
#'  "negative").}
#'  \item{nhg}{Nottingham histological grade.}
#'  \item{pam50_subtype}{PAM50 subtype.}
#'  \item{overall_survival_days}{Days of overall survival.}
#'  \item{overall_survival_event}{Overall survival event (1 for "deceased", 0
#'  for "alive").}
#' }
#' @usage data(samplesBC)
#'
"samplesBC"

#' A subset of genes involved in early estrogen response
#'
#' This vector contains a subset of genes involved in early estrogen response.
#'
#' @format A character vector
#' @usage data(genesER)
"genesER"


R/DCMat.R
#' Differential Co-expression
#'
#' Calculate global differential co-expression network.
#'
#' This function calculates a global differential co-expression network between
#' two conditions using a z-score-based method. This method calculates z-scores
#' through the application of Fisher transformation to the correlation
#' coefficients and a subsequent z-test.
#'
#' @param exprMat an expression matrix. Rows represent genes and
#' columns represent samples.
#' @param condition a vector that specifies the condition of each sample.
#' The vector should contain only two conditions. The order of the vector
#' should match the column order of the expression matrix.
#' @param corMethod a character string indicating the correlation coefficient
#' to use. Can be either "spearman" (default) or "pearson".
#'
#' @return a matrix of z-scores representing the global differential
#' co-expression network between the two input conditions.
#'
#' @examples
#' # obtain expression data
#' data(exprBC)
#' # obtain sample information
#' data(samplesBC)
#'
#' # calculate differential correlation matrix
#' dcMat <- DCMat(exprMat = exprBC,
#'                condition = samplesBC$er_status,
#'                corMethod = "spearman")
#'
#' @author Lanying Wei
#'
#' @seealso \code{\link{DCPlot}}
#' @seealso \code{\link{eigenCentrality}}
#'
#' @export

DCMat <- function(exprMat,
                  condition,
                  corMethod=c("spearman", "pearson")) {

    stopifnot(!missing(exprMat),
              !missing(condition),
              length(condition) == ncol(exprMat))

    corMethod <- match.arg(corMethod)
    condition <- factor(condition)

    if (length(levels(condition)) != 2) {
        stop("There should be two levels of condition.")
    }

    message("Calculating differential coexpression matrix for ",
            levels(condition)[2],
            " (n=",
            sum(condition %in% levels(condition)[2]),
            ") ",
            "versus ", levels(condition)[1],
            " (n=",
            sum(condition %in% levels(condition)[1]),
            ") ...")

    exprMat1 <- exprMat[, condition %in% levels(condition)[1], drop = FALSE]
    exprMat2 <- exprMat[, condition %in% levels(condition)[2], drop = FALSE]

    const <- 1
    if (corMethod %in% 'spearman') {
        const <- 1.06
        message("Use Spearman correlation.")
    } else {
        message("Use Pearson correlation.")
    }

    message("Calculating correlation matrix for the first condition...")
    r1 <- getCor(t(exprMat1), corMethod)
    message("Calculating correlation matrix for the second condition...")
    r2 <- getCor(t(exprMat2), corMethod)
    message("Fisher transformation...")
    z1 <- atanh(r1)
    z2 <- atanh(r2)
    z <- (z2-z1)/sqrt(const/(sum(condition %in% levels(condition)[2])-3)+
                          const/(sum(condition %in% levels(condition)[1])-3))
    z[is.nan(z)] <- 0
    z[is.infinite(z)] <- 0
    diag(z) <- 0

    message("Done.")

    return(z)
}

getCor <- function(exprMat,
                   corMethod=c("spearman", "pearson")) {

    corMethod <- match.arg(corMethod)

    if (corMethod %in% "spearman") {
        exprMat <- apply(exprMat, 2, rank)
    }

    design <- matrix(data = 1,
                     nrow = nrow(exprMat),
                     ncol = 1)
    QR <- qr(design)
    E <- qr.qty(QR, exprMat)
    s2 <- colMeans(E[-1, ]^2)
    U <- t(t(E[-1, ])/sqrt(s2))
    c <- crossprod(U)/nrow(U)
    diag(c) <- 1

    return(c)
}




R/DCPlot.R
#' Differential Co-expression Network Plotting
#'
#' Plot differential co-expression network for a given set of genes.
#'
#' This function plots correlation networks for a given set of genes separately
#' for two conditions. Edge weights represent the correlation coefficient
#' between two genes. The users can set various thresholds and colors to focus
#' on key changes in the network and understand the change in associations
#' between gene pairs under different conditions.
#'
#'
#' @param exprMat an expression matrix. Rows represent genes and
#' columns represent samples.
#' @param condition a vector that specifies the condition of each sample.
#' The vector should contain only two conditions. The order of the vector
#' should match the column order of the expression matrix.
#' @param genes a character vector of genes to plot.
#' @param dcMat a matrix of z-scores representing differential associations
#' of gene expression between two conditions. See {\link{DCMat}} for more
#' details. This matrix, along with posDCCut and negDCCut, are used to remove
#' edges between gene pairs that show low level of differential co-expression.
#' @param corMethod a character string indicating the correlation coefficient
#' to use. Can be either "spearman" (default) or "pearson".
#' @param posDCCut a positive numeric value that serves as positive
#' threshold for DCMat to hide unimportant edges. Gene pairs with DCMat values
#' below posDCCut and above negDCCut are hidden.
#' @param negDCCut a negative numeric value that serves as negative
#' threshold for DCMat to hide unimportant edges. Gene pairs with DCMat values
#' below posDCCut and above negDCCut are hidden.
#' @param minimum a numeric value between 0 and 1. Edges with absolute
#' weights less than this value are not shown.
#' @param maximum a numeric value between 0 and 1. The highest weight to scale
#' the edge widths to.
#' @param cut a numeric value between 0 and 1. Cut the scaling of edges in
#' width and color saturation. Edges with absolute weights over this value
#' have the strongest color intensity and become wider the stronger they are,
#' and edges with absolute weights under this value have the smallest width
#' and become vaguer the weaker the weight. If this is set to 0, no cutoff is
#' used and all edges vary in width and color.
#' @param posCol color of positive edges. Can be a vector of two to indicate
#' color of edges under 'cut' value and color of edges over 'cut' value.
#' @param negCol color of negative edges. Can be a vector of two to indicate
#' color of edges under 'cut' value and color of edges over 'cut' value.
#' @param layout the layout of the graph, can be "spring" or "circle". "spring"
#' gives a force embedded layout; "circle" places all nodes in a single
#' circle.
#' @param repulsion Scalar on the default repulse radius in the spring layout.
#' Defaults to 1. Setting this argument to lower values (e.g., 0.5) will cause
#' nodes in the spring layout to repulse each other less. This is especially
#' useful if a few unconnected nodes cause the giant component to visually
#' be clustered too much in the same place.
#' @param labelCex numeric scalar on the label size.
#' @param labelCol Character containing the color of the labels.
#' @param geneList1 a character vector of a list of genes to be colored.
#' Genes in this vector are labelled with color 'labelCol1'.
#' @param geneList2 a character vector of a list of genes to be colored.
#' Genes in this vector are labelled with color 'labelCol2'.
#' @param labelCol1 color of genes in geneList1.
#' @param labelCol2 color of genes in geneList2.
#' @param cond1Title plot title for the first condition.
#' @param cond2Title plot title for the second condition.
#' @param ... other parameters to be passed to \code{\link{qgraph}}.
#'
#' @return a list of two "qgraph" objects, each corresponds to one of the
#' correlation plots of the two input conditions. See \code{\link{qgraph}}
#' for more details.
#'
#' @examples
#'
#' # get expression data
#' data(exprBC)
#' # get sample information
#' data(samplesBC)
#' # get example genes
#' data(genesER)
#'
#' # calculate differential correlation matrix
#' dcMat <- DCMat(exprMat = exprBC,
#'                condition = samplesBC$er_status,
#'                corMethod = "spearman")
#'
#' # plot differential co-expression network using example genes
#' p <- DCPlot(exprMat = exprBC,
#'             condition = samplesBC$er_status,
#'             genes = genesER,
#'             dcMat = dcMat,
#'             posDCCut = 5,
#'             negDCCut = -5,
#'             corMethod = "spearman",
#'             minimum = 0.2)
#'
#' @author Lanying Wei
#'
#' @seealso \code{\link{DCMat}}
#' @seealso \code{\link{eigenCentrality}}
#'
#' @importFrom qgraph qgraph averageLayout
#' @importFrom graphics par
#'
#' @export


DCPlot <- function(exprMat,
                   condition,
                   genes,
                   dcMat,
                   corMethod=c("spearman", "pearson"),
                   posDCCut=3,
                   negDCCut=-posDCCut,
                   minimum=0.2,
                   maximum=1,
                   cut=0.8,
                   posCol=c("#4361ee", "#3f37c9"),
                   negCol=c("#ff5a5f", "#d90429"),
                   layout=c("spring", "circle"),
                   repulsion=1,
                   labelCex=1,
                   labelCol="#6c757d",
                   geneList1=NULL,
                   geneList2=NULL,
                   labelCol1="#e63946",
                   labelCol2="#072ac8",
                   cond1Title=NULL,
                   cond2Title=NULL,
                   ...) {

    stopifnot(!missing(exprMat),
              !missing(condition),
              !missing(genes),
              !missing(dcMat),
              length(condition) == ncol(exprMat),
              all(colnames(dcMat) == rownames(dcMat)))

    condition <- factor(condition)
    if (length(levels(condition)) != 2) {
        stop("There should be two levels of condition.")
    }
    if (!all(genes %in% rownames(exprMat))) {
        stop("Gene ",
             paste(genes[!genes %in% rownames(exprMat)], collapse = " "),
             " not in exprMat.")
    }
    if (!all(genes %in% rownames(dcMat))) {
        stop("Gene ",
             paste(genes[!genes %in% rownames(dcMat)], collapse = " "),
             " not in dcMat." )
    }

    corMethod <- match.arg(corMethod)
    layout <- match.arg(layout)

    corMat1 <-
        getCor(t(exprMat[genes, condition %in% levels(condition)[1]]),
               corMethod)
    corMat2 <-
        getCor(t(exprMat[genes, condition %in% levels(condition)[2]]),
               corMethod)

    corMat1[which(dcMat[genes, genes] > negDCCut &
                      dcMat[genes, genes] < posDCCut, arr.ind = TRUE)] <- 0
    corMat2[which(dcMat[genes, genes] > negDCCut &
                      dcMat[genes, genes] < posDCCut, arr.ind = TRUE)] <- 0


    corMat1[abs(corMat1) < minimum] <- 0
    corMat2[abs(corMat2) < minimum] <- 0
    genes2remove <- names(which(colSums(corMat1) == 0 & colSums(corMat2 == 0)))

    corMat1 <- corMat1[!rownames(corMat1) %in% genes2remove,
                       !colnames(corMat1) %in% genes2remove]

    corMat2 <- corMat2[!rownames(corMat2) %in% genes2remove,
                       !colnames(corMat2) %in% genes2remove]

    genes <- colnames(corMat1)

    corMat1 <-
        getCor(t(exprMat[genes, condition %in% levels(condition)[1]]),
               corMethod)
    corMat2 <-
        getCor(t(exprMat[genes, condition %in% levels(condition)[2]]),
               corMethod)

    corMat1[which(dcMat[genes, genes] > negDCCut &
                      dcMat[genes, genes] < posDCCut, arr.ind = TRUE)] <- 0
    corMat2[which(dcMat[genes, genes] > negDCCut &
                      dcMat[genes, genes] < posDCCut, arr.ind = TRUE)] <- 0


    L <- averageLayout(corMat1, corMat2,
                       repulsion = repulsion, layout = layout)

    labelColor <- rep(labelCol, ncol(corMat1))
    labelColor[colnames(corMat1) %in% geneList1] <- labelCol1
    labelColor[colnames(corMat1) %in% geneList2] <- labelCol2


    if (is.null(cond1Title)) {
        cond1Title <- levels(condition)[1]
    }

    if (is.null(cond2Title)) {
        cond2Title <- levels(condition)[2]
    }
    oldMfrow <- par()$mfrow
    par(mfrow = c(1, 2))

    p1 <- qgraph(corMat1, layout = L,
                 labels = colnames(corMat1),
                 label.cex = labelCex,
                 label.scale = FALSE,
                 minimum = minimum,
                 maximum = maximum,
                 cut = cut,
                 vsize = 0,
                 label.color = labelColor,
                 posCol = posCol,
                 negCol = negCol,
                 title = cond1Title,
                 ...)


    p2 <- qgraph(corMat2, layout = L,
                 labels = colnames(corMat1),
                 label.cex = labelCex,
                 label.scale = FALSE,
                 minimum = minimum,
                 maximum = maximum,
                 cut = cut,
                 vsize = 0,
                 label.color = labelColor,
                 posCol = posCol,
                 negCol = negCol,
                 title = cond2Title,
                 ...)

    par(mfrow = oldMfrow)

    res <- list(p1, p2)
    names(res) <- c(cond1Title, cond2Title)
    return(res)
}



R/cosinet.R
#' @details
#'
#' Cosinet is an R package that aims to bridge the gap between differential
#' co-expression network analysis and precision medicine by accurately
#' quantifying the degree of rewiring of co-expression networks at
#' the level of individual samples. Differential co-expression network analysis
#' is a valuable approach for understanding the underlying mechanisms of disease
#' and informing the development of personalized therapies, but current
#' approaches typically measure an average differential network across multiple
#' samples, making it difficult to apply in the field of precision medicine.
#' Cosinet utilizes gene expression data to determine the degree of similarity
#' between the gene co-expression patterns of a given sample and reference
#' conditions within the context of a function-specific differential
#' co-expression network. The package employs a combination of techniques
#' such as differential co-expression analysis, network centrality calculation,
#' gene set enrichment analysis (GSEA), and a novel statistic that measures the
#' differences in statistical independence of a gene pair between two
#' conditions for a single sample. The general workflow is as follows:
#'
#'
#' @section Construct global differential co-expression network:
#' \itemize{
#' \item{\code{\link{matPreprocess}}}: Preprocess expression matrix for
#'     co-expression analysis. There should be two conditions of samples
#'     for comparison.
#' \item{\code{\link{DCMat}}}: Calculate differential co-expression between
#' the two conditions using a z-score based method.}
#' @section Calculate node centralities in differential co-expression network:
#' \itemize{
#' \item{\code{\link{eigenCentrality}}}: Calculate eigenvector centrality
#' using differential co-expression matrix.
#' \item{\code{\link{rankPlot}}}: Visualize the score and rank of gene
#' centrality.
#' }
#' @section Identify function-specific differential co-expression sub-networks:
#' Then the resulting gene centrality scores can be used as input for GSEA
#' using for example \code{GSEA} function from \code{clusterProfiler}
#' package. The core-enrichment genes of a relevant gene set can be used to
#' compute Cosinet scores.
#' @section Compute Cosinet scores based on differential co-expression sub-network:
#' \itemize{
#' \item{\code{\link{DCPlot}}}: Plot differential co-expression network for
#' a given set of genes.
#' \item{\code{\link{getCosi}}}: Calculate Cosinet score for a given set of
#' genes.
#' }
#' The final Cosinet score for each sample represents a measure of how
#' closely the gene co-expression patterns of that sample match the
#' patterns observed in the reference conditions within the differential
#' sub-network. A lower Cosinet score indicates that the gene co-expression
#' patterns of the given sample are more similar to those of the first
#' condition, while a higher Cosinet score indicates that they are more
#' similar to those of the second condition. Downstream analysis can then
#' be performed to further investigate the differences.
#'
#' @author Lanying Wei
#'
#' @seealso \code{\link{matPreprocess}}
#' @seealso \code{\link{DCMat}}
#' @seealso \code{\link{eigenCentrality}}
#' @seealso \code{\link{rankPlot}}
#' @seealso \code{\link{DCPlot}}
#' @seealso \code{\link{getCosi}}
#'
"_PACKAGE"


R/rankPlot.R
#' Rank Plot
#'
#' Draw the score and rank of genes on a scatter plot.
#'
#' @param geneScore a numeric vector of scores to rank, with gene as names.
#' @param genesToLabel a character vector specifying the genes to be labeled.
#' @param nLabel an integer specifying the number of top scored genes to be
#' labeled.
#' @param nIQR a positive numeric value indicating the fold of IQR to be used as
#' cutoff to color genes. Genes with scores > Q3 + nIQR * (Q3 - Q1) are colored
#' red.
#' @param labelSize a positive numeric value specifying the size of the label.
#' @param main a character string representing the plot title.
#'
#' @return A \code{ggplot} scatter plot object showing the ranked gene scores.
#'
#' @author Lanying Wei
#'
#' @examples
#' # get expression data
#' data(exprBC)
#' # get sample information
#' data(samplesBC)
#'
#' # calculate differential correlation matrix between ER status
#' dcMat <- DCMat(exprMat = exprBC,
#'                condition = samplesBC$er_status,
#'                corMethod = "spearman")
#'
#' # calculate gene centrality base on differential correlation matrix
#' geneCentrality <- eigenCentrality(mat = dcMat)
#'
#' # plot gene centrality rank plot
#' rankPlot(geneCentrality)
#'
#' @import ggplot2
#' @importFrom ggrepel geom_text_repel
#' @importFrom stats quantile
#'
#' @seealso \code{\link{eigenCentrality}}
#'
#'
#' @export

rankPlot <- function(geneScore,
                     genesToLabel=NULL,
                     nLabel=10,
                     nIQR=1.5,
                     labelSize=3,
                     main=NULL) {

    stopifnot(!missing(geneScore))

    Q1 <- quantile(geneScore, 0.25)
    Q3 <- quantile(geneScore, 0.75)
    cutoff <- Q3 + nIQR * (Q3 - Q1)

    data <- data.frame(Gene = names(geneScore),
                       Score = geneScore)
    data$Rank <- rank(-data$Score)
    data$TopGenes <- "No"
    data$TopGenes[data$Score > cutoff] <- "Yes"

    myColor <- c("No" = "#c0c0c0",
                 "Yes" = "#e63946")

    idx <- (data$Rank <= nLabel) | (data$Gene %in% genesToLabel)
    p <- ggplot(data, aes_string(x = "Rank",
                                 y = "Score",
                                 color = "TopGenes")) +
        geom_point(size = 0.5)
    if(sum(idx) > 0) {
        p <- p + ggrepel::geom_text_repel(aes_string(label = "Gene"),
                                          data = data[idx,],
                                          size = labelSize,
                                          max.overlaps = Inf)
    }

    p <- p + scale_color_manual(values = myColor) +
        scale_fill_manual(values = myColor) +
        labs(x = "Rank",
             y = "Score",
             title = main) +
        theme_bw(base_size = 14) +
        theme(plot.title = element_text(hjust = 0.5),
              legend.position = "none")
    return(p)
}


R/getCosi.R
#' Cosinet Score
#'
#' Calculate Cosinet scores with a given set of genes.
#'
#' Calculate Cosinet scores with a given set of genes. Cosinet uses gene
#' expression data to determine the degree of similarity between the
#' gene co-expression patterns of a given sample and reference conditions
#' in the context of a function-specific differential co-expression network.
#' Cosinet scores quantifies the degree of rewiring of co-expression network at
#' the level of individual samples.
#'
#'
#' @param exprMat an expression matrix. Rows represent genes and
#' columns represent samples.
#' @param condition a vector that specifies the condition of each sample.
#' The vector should contain only two conditions. The order of the vector
#' should match the column order of the expression matrix.
#' @param genes a character vector of genes to be evaluated. These genes can be
#' the genes involved in a function-specific differential co-expression network.
#' Gene pairs generated from this gene list that pass the thresholds for
#' dcMat are used to calculate Cosinet scores.
#' @param sampleNames a character vector of sample names to calculate the
#' Cosinet scores for.
#' @param bin a numeric value between 0 and 1 that represents the fraction of
#' the bin size relative to the binRange length. Binned regions are used to
#' estimate gene pair dependency at a local region around an individual sample.
#' @param binRange a numeric value between 0 and 1 that represents the fraction
#' of the middle expression range relative to the full expression range. The
#' length of binRange is used as the base to generate the bin length. When
#' binRange is set to 1, the range between the maximum expression value and
#' the minimum expression value is used as the base. If users are concerned
#' about extreme data having a large impact on the bin size setting, a smaller
#' value of binRange, for example 0.95, can be set so that the expression range
#' between the 2.5% quantile and the 97.5% quantile is used as the base. A
#' binRange value less than 0.9 is not recommended, as this will exclude up to
#' 10% of the samples from the bin size calculation.
#' @param dcMat a matrix of z-scores representing differential associations
#' of gene expression between two conditions. See {\link{DCMat}} for more
#' details. This matrix, along with posDCCut and negDCCut, are used to 1. remove
#' gene pairs that show low level of differential co-expression, so that only
#' important changes in the network are considered in the Cosinet score
#' calculation; and 2. used as weights when calculating Cosinet scores.
#' @param posDCCut a positive numeric value that serves as positive
#' threshold for dcMat. Gene pairs with dcMat values
#' below posDCCut and above negDCCut are not used for the calculation.
#' @param negDCCut a negative numeric value that serves as negative
#' threshold for dcMat. Gene pairs with dcMat values
#' below posDCCut and above negDCCut are not used for the calculation.
#' @param weighted logical indicator for weighting each gene pair by the
#' absolute value of corresponding dcMat score during Cosinet score calculation.
#' @param threads number of threads used for calculation.
#' @return a matrix of the final Cosinet scores and the individual scores for
#' each gene pair before averaging. A lower Cosinet score indicates that the
#' gene co-expression patterns of the given sample are more similar to those of
#' the first condition, while a higher Cosinet score indicates that they are
#' more similar to those of the second condition.
#'
#'
#' @author Lanying Wei
#'
#' @seealso \code{\link{DCMat}}
#' @seealso \code{\link{eigenCentrality}}
#'
#' @importFrom foreach foreach %dopar%
#' @importFrom parallel makeCluster stopCluster
#' @importFrom doParallel registerDoParallel
#' @importFrom stats p.adjust t.test
#' @importFrom utils combn
#'
#'
#' @examples
#'
#' # get expression data
#' data(exprBC)
#' # get sample information
#' data(samplesBC)
#' # get example genes
#' data(genesER)
#'
#' # calculate differential correlation matrix
#' dcMat <- DCMat(exprMat = exprBC,
#'                condition = samplesBC$er_status,
#'                corMethod = "spearman")
#'
#' # calculate Cosinet scores with example genes
#' scoreMat <- getCosi(exprMat = exprBC,
#'                     condition = samplesBC$er_status,
#'                     genes = genesER,
#'                     bin = 0.1,
#'                     dcMat = dcMat,
#'                     posDCCut = 3,
#'                     negDCCut = -3,
#'                     threads = 2)
#'
#'
#' @export

getCosi <- function(exprMat,
                    condition,
                    genes,
                    sampleNames=colnames(exprMat),
                    bin=0.1,
                    binRange=1,
                    dcMat=NULL,
                    posDCCut=3,
                    negDCCut=-posDCCut,
                    weighted=TRUE,
                    threads=2) {

    stopifnot(!missing(exprMat),
              !missing(condition),
              !missing(genes),
              length(condition) == ncol(exprMat),
              bin > 0 & bin < 1)

    condition <- factor(condition)

    if (length(levels(condition)) != 2) {
        stop("There should be two levels of condition.")
    }

    if (!all(genes %in% rownames(exprMat))) {
        stop("Gene ",
             paste(genes[!genes %in% rownames(exprMat)], collapse = " "),
             " not in exprMat.")
    }


    if (!all(sampleNames %in% colnames(exprMat))) {
        stop("Sample ",
             paste(sampleNames[!sampleNames %in% colnames(exprMat)],
                   collapse = " "),
             " not in exprMat.")
    }

    condition <- factor(condition)
    genePairs <- as.data.frame(t(combn(genes, 2)))
    weights <- apply(genePairs, 1, function(x) dcMat[x[1], x[2]])

    if (!is.null(dcMat) & !is.null(posDCCut) & !is.null(negDCCut)) {
        if (any(weights > posDCCut | weights < negDCCut)) {
            nGenePairsAll <- nrow(genePairs)
            genePairs <- genePairs[weights > posDCCut | weights < negDCCut, ,
                                   drop = FALSE]
            weights <- weights[weights > posDCCut | weights < negDCCut]
            message(nGenePairsAll - nrow(genePairs),
                    " gene pairs that failed to meet the dcMat thresholds",
                    " were removed from the Cosinet score calculation.")
        } else {
            stop("No gene pair meets the given dcMat thresholds.")
        }
    }

    deltaScores <- deltaScore(exprMat = exprMat,
                              genePairs = genePairs,
                              condition = condition,
                              sampleNames = sampleNames,
                              bin = bin,
                              binRange = binRange,
                              threads = threads)

    if (!is.null(dcMat) & weighted == TRUE) {
        message("Weight each gene pair by absolute dcMat value.")
        deltaScores <- deltaScores * abs(weights)
    }

    deltaScores <- t(deltaScores)
    Cosinet <- apply(deltaScores, 1, function(x) mean(x, na.rm = TRUE))

    finalRes <- cbind(Cosinet, deltaScores)
    finalRes <- as.data.frame(finalRes)

    message(ncol(deltaScores),
            " gene pairs contributed to the final Cosinet score calculation.")

    return(finalRes)
}

deltaScore <- function(exprMat,
                       genePairs,
                       condition,
                       sampleNames=colnames(exprMat),
                       bin=0.1,
                       binRange = binRange,
                       threads=2) {

    condition <- factor(condition)
    genes2Use <- unique(c(genePairs[, 1], genePairs[, 2]))
    exprMat1 <- exprMat[genes2Use, condition %in% levels(condition)[1]]
    exprMat2 <- exprMat[genes2Use, condition %in% levels(condition)[2]]

    cl <- makeCluster(threads)
    registerDoParallel(cl)

    message("Performing calculation using ", threads, " threads...")

    nSamples <- ceiling(length(sampleNames) / threads)
    nTimes <- min(threads, ceiling(length(sampleNames) / nSamples))

    i <- 1
    res <- foreach(i = seq_len(nTimes),
                   .combine = "c",
                   .inorder = TRUE,
                   .verbose = FALSE,
                   .export = c("rhoScore")
    ) %dopar% {
        start <- (i - 1) * nSamples + 1
        end <- min(i * nSamples, length(sampleNames))
        deltaList <- list()
        for (sampleName in sampleNames[start:end]) {
            if (!sampleName %in% colnames(exprMat1)) {
                m1 <- cbind(exprMat1, exprMat[genes2Use, sampleName, drop = FALSE])
                colnames(m1)[ncol(m1)] <- sampleName
            } else {
                m1 <- exprMat1
            }
            if (!sampleName %in% colnames(exprMat2)) {
                m2 <- cbind(exprMat2, exprMat[genes2Use, sampleName, drop = FALSE])
                colnames(m2)[ncol(m2)] <- sampleName
            } else {
                m2 <- exprMat2
            }
            rho1 <- rhoScore(m1, sampleName, genePairs,
                             bin = bin, binRange = binRange)
            rho2 <- rhoScore(m2, sampleName, genePairs,
                             bin = bin, binRange = binRange)
            rho1[rho1 < 0] <- 0
            rho2[rho2 < 0] <- 0
            delta <- rho2 - rho1

            deltaList[[sampleName]] <- delta
        }
        return(deltaList)
    }
    res <- as.data.frame(res)
    rownames(res) <- paste(genePairs[, 1], genePairs[, 2], sep = "_")

    stopCluster(cl)
    return(res)
}

rhoScore <- function(m,
                     sampleName,
                     genePairs,
                     bin=0.1,
                     binRange=1) {
    n <- ncol(m)
    sideRange <- (1 - binRange) / 2
    lowerExp <- apply(m[, !colnames(m) %in% sampleName], 1,
                      function(x) quantile(x = x,
                                           probs = sideRange,
                                           na.rm = TRUE,
                                           names = FALSE))
    upperExp <- apply(m[, !colnames(m) %in% sampleName], 1,
                      function(x) quantile(x = x,
                                           probs = 1 - sideRange,
                                           na.rm = TRUE,
                                           names = FALSE))
    binVal <- (upperExp - lowerExp) * bin * 0.5
    diffExp <- abs(m - m[, sampleName])
    diffExp[, sampleName] <- Inf
    binMat <- diffExp <= binVal
    binMat <- t(binMat * 1)
    rhoScores <- mapply(function(gene1, gene2) {
        x <- binMat[, gene1]
        y <- binMat[, gene2]
        (sum(x & y) / n - sum(x) / n * sum(y) / n) * 100
    }, genePairs[, 1], genePairs[, 2], SIMPLIFY = TRUE, USE.NAMES = FALSE)

    return(rhoScores)
}
